Index: app/app-Info.plist
===================================================================
--- app/app-Info.plist	(revision 220934)
+++ app/app-Info.plist	(working copy)
@@ -220,13 +220,13 @@
 	<key>CFBundleExecutable</key>
 	<string>${EXECUTABLE_NAME}</string>
 	<key>CFBundleIconFile</key>
-	<string>app.icns</string>
+	<string>hex.icns</string>
 	<key>CFBundleIdentifier</key>
 	<string>${CHROMIUM_BUNDLE_ID}</string>
 	<key>CFBundleInfoDictionaryVersion</key>
 	<string>6.0</string>
 	<key>CFBundleName</key>
-	<string>${CHROMIUM_SHORT_NAME}</string>
+	<string>HeXium</string>
 	<key>CFBundlePackageType</key>
 	<string>APPL</string>
 	<key>CFBundleSignature</key>
Index: app/chrome_dll.rc
===================================================================
--- app/chrome_dll.rc	(revision 220934)
+++ app/chrome_dll.rc	(working copy)
@@ -180,7 +180,8 @@
 IDR_MAINFRAME       ICON                        "theme\google_chrome\chrome.ico"
 IDR_SXS             ICON                        "theme\google_chrome\chrome_sxs.ico"
 #else
-IDR_MAINFRAME       ICON                        "theme\chromium\chromium.ico"
+//IDR_MAINFRAME       ICON                        "theme\chromium\chromium.ico"
+IDR_MAINFRAME           ICON                    "../../hex/client/res/hex.ico"
 #endif
 
 // We include these resources because all ICON types need to be in the
Index: app/chrome_exe.rc
===================================================================
--- app/chrome_exe.rc	(revision 220934)
+++ app/chrome_exe.rc	(working copy)
@@ -44,7 +44,8 @@
 // identify it.
 IDR_SXS                 ICON                    "theme\google_chrome\chrome_sxs.ico"
 #else
-IDR_MAINFRAME           ICON                    "theme\chromium\chromium.ico"
+//IDR_MAINFRAME           ICON                    "theme\chromium\chromium.ico"
+IDR_MAINFRAME           ICON                    "../../hex\client/res/hex.ico"
 #endif
 
 // Icon for the App Lancher. Named with a leading X to keep the name
Index: app/chromium_strings.grd
===================================================================
--- app/chromium_strings.grd	(revision 220934)
+++ app/chromium_strings.grd	(working copy)
@@ -206,16 +206,16 @@
   <release seq="1" allow_pseudo="false">
     <messages fallback_to_english="true">
       <message name="IDS_PRODUCT_NAME" desc="The Chrome application name">
-        Chromium
+        HeXium
       </message>
       <message name="IDS_SHORT_PRODUCT_NAME" desc="The Chrome application short name.">
-        Chromium
+        HeXium
       </message>
       <message name="IDS_SHORT_PRODUCT_NAME_LOWER" desc="The Chrome application short name all lowercase">
-        chromium
+        hexium
       </message>
       <message name="IDS_SXS_SHORTCUT_NAME" desc="The Chrome short cut name for SxS channel">
-        Chromium
+        HeXium
       </message>
       <message name="IDS_PRODUCT_DESCRIPTION" desc="Browser description">
         Chromium is a web browser that runs webpages and applications with lightning speed. It's fast, stable, and easy to use. Browse the web more safely with malware and phishing protection built into Chromium.
@@ -305,14 +305,14 @@
         You attempted to reach &lt;strong&gt;<ph name="DOMAIN">$1<ex>paypal.com</ex></ph>&lt;/strong&gt;, but the certificate that the server presented contains errors. Chromium cannot use a certificate with errors and cannot validate the identity of the site that you have attempted to connect to.
       </message>
       <message name="IDS_TASK_MANAGER_TITLE" desc="The title of the Task Manager window">
-        Task Manager - Chromium
+        Task Manager - HeXium
       </message>
       <message name="IDS_SESSION_CRASHED_VIEW_MESSAGE" desc="Message shown when the last session didn't exit cleanly.">
         Chromium didn't shut down correctly. To reopen the pages you had open, click Restore.
       </message>
       <!-- Browser Window Title Format -->
       <message name="IDS_BROWSER_WINDOW_TITLE_FORMAT" desc="The format for titles displayed in tabs and popup windows">
-        <ph name="PAGE_TITLE">$1<ex>Google</ex></ph> - Chromium
+        <ph name="PAGE_TITLE">$1</ph> - HeXium
       </message>
       <if expr="not pp_ifdef('chromeos')">
         <message name="IDS_GET_HELP_USING_CHROME" desc="Text of the button which takes the user to the Chrome help page.">
@@ -328,7 +328,7 @@
         The Chromium Authors
       </message>
       <message name="IDS_ABOUT_VERSION_COPYRIGHT" desc="Copyright information on the about pages">
-        Copyright 2013 The Chromium Authors. All rights reserved.
+        Copyright 2013 The Chromium Authors. All rights reserved. Portions Copyright (c) 2013 NetEase Youdao Inc. and other heX contributors.
       </message>
       <message name="IDS_ABOUT_VERSION_LICENSE" desc="The label below the copyright message, containing the URLs.">
         Chromium is made possible by the <ph name="BEGIN_LINK_CHROMIUM">&lt;a target="_blank" href="$1"&gt;</ph>Chromium<ph name="END_LINK_CHROMIUM">&lt;/a&gt;</ph> open source project and other <ph name="BEGIN_LINK_OSS">&lt;a target="_blank" href="$2"&gt;</ph>open source software<ph name="END_LINK_OSS">&lt;/a&gt;</ph>.
@@ -342,7 +342,7 @@
         Not used in Chromium. Placeholder to keep resource maps in sync. It expects one argument: $1.
       </message>
       <message name="IDS_ACCNAME_APP" desc="The accessible name for the app menu.">
-        Chromium
+        HeXium
       </message>
       <message name="IDS_ACCNAME_TOOLBAR" desc="The accessible name for the application's toolbar.">
         Chromium Toolbar
Index: app/generated_resources.grd
===================================================================
--- app/generated_resources.grd	(revision 220934)
+++ app/generated_resources.grd	(working copy)
@@ -9502,7 +9502,7 @@
       </if>
       <if expr="not pp_ifdef('chromeos')">
         <message name="IDS_ABOUT_PRODUCT_DESCRIPTION" desc="The text label describing the Chrome browser">
-          A web browser built for speed, simplicity, and security
+          A web browser based on Chromium combining Node.js
         </message>
       </if>
       <message name="IDS_ABOUT_PRODUCT_VERSION" desc="The text label describing the version of the browser">
Index: app/resources/generated_resources_en-GB.xtb
===================================================================
--- app/resources/generated_resources_en-GB.xtb	(revision 220934)
+++ app/resources/generated_resources_en-GB.xtb	(working copy)
@@ -566,7 +566,7 @@
 <translation id="7800304661137206267">The connection is encrypted using <ph name="CIPHER"/>, with <ph name="MAC"/> for message authentication and <ph name="KX"/> as the key exchange mechanism.</translation>
 <translation id="350893259022641366">Per-tile Painting</translation>
 <translation id="7706319470528945664">Portuguese keyboard</translation>
-<translation id="7331786426925973633">A web browser built for speed, simplicity, and security</translation>
+<translation id="7331786426925973633">A web browser based on Chromium combining Node.js</translation>
 <translation id="5584537427775243893">Importing</translation>
 <translation id="9128870381267983090">Connect to network</translation>
 <translation id="4779735050989188948">Manipulate settings that control websites' access to cookies, JavaScript, and plug-ins</translation>
Index: app/resources/generated_resources_zh-CN.xtb
===================================================================
--- app/resources/generated_resources_zh-CN.xtb	(revision 220934)
+++ app/resources/generated_resources_zh-CN.xtb	(working copy)
@@ -544,7 +544,7 @@
 <translation id="7800304661137206267">该连接是使用 <ph name="CIPHER"/> 进行加密的，同时使用 <ph name="MAC"/> 进行讯息身份验证并使用 <ph name="KX"/> 作为密钥交换机制。</translation>
 <translation id="350893259022641366">逐块绘图</translation>
 <translation id="7706319470528945664">葡萄牙语键盘</translation>
-<translation id="7331786426925973633">追求速度、简约和安全的网络浏览器</translation>
+<translation id="7331786426925973633">基于 Chromium、融合了 Node.js 的网络浏览器</translation>
 <translation id="5584537427775243893">正在导入</translation>
 <translation id="9128870381267983090">连接到网络</translation>
 <translation id="4779735050989188948">管理用于控制网站对 Cookie、JavaScript 和插件的访问权限的设置</translation>
Index: app/resources/generated_resources_zh-TW.xtb
===================================================================
--- app/resources/generated_resources_zh-TW.xtb	(revision 220934)
+++ app/resources/generated_resources_zh-TW.xtb	(working copy)
@@ -556,7 +556,7 @@
 <translation id="7800304661137206267">連線採用 <ph name="CIPHER"/> 加密，並設有 <ph name="MAC"/> 訊息驗證及 <ph name="KX"/> 金鑰交換機制。</translation>
 <translation id="350893259022641366">Per-tile 繪圖</translation>
 <translation id="7706319470528945664">葡萄牙文鍵盤</translation>
-<translation id="7331786426925973633">簡單、快速而又安全的網路瀏覽器</translation>
+<translation id="7331786426925973633">包含 Node.js 的 Chromium 的網路瀏覽器</translation>
 <translation id="5584537427775243893">匯入中</translation>
 <translation id="9128870381267983090">連線至網路</translation>
 <translation id="4779735050989188948">操控決定網站能否存取 Cookie、JavaScript 和外掛程式的設定</translation>
Index: app/resources/locale_settings.grd
===================================================================
--- app/resources/locale_settings.grd	(revision 220934)
+++ app/resources/locale_settings.grd	(working copy)
@@ -347,8 +347,9 @@
 
       <!-- The URL for Google Chrome welcome page. This is used for the pre -->
       <!-- populated thumbnails. -->
+      <!-- For heXium -->
       <message name="IDS_CHROME_WELCOME_URL" translateable="false">
-        http://www.google.com/chrome/intl/[GRITLANGCODE]/welcome.html
+        http://hex.youdao.com/hexium/welcome.html
       </message>
       <if expr="is_android or is_ios">
         <message name="IDS_MOBILE_WELCOME_URL" translateable="false">
Index: app/theme/chrome_unscaled_resources.grd
===================================================================
--- app/theme/chrome_unscaled_resources.grd	(revision 220934)
+++ app/theme/chrome_unscaled_resources.grd	(working copy)
@@ -33,18 +33,18 @@
         <include name="IDR_WIZARD_ICON_RTL" file="google_chrome/wizard_icon_rtl.png" type="BINDATA" />
       </if>
       <if expr="not pp_ifdef('_google_chrome')">
-        <include name="IDR_PRODUCT_LOGO_64" file="chromium/product_logo_64.png" type="BINDATA" />
-        <include name="IDR_PRODUCT_LOGO_128" file="chromium/product_logo_128.png" type="BINDATA" />
+        <include name="IDR_PRODUCT_LOGO_64" file="../../../hex/browser/resources/hex_logo_64.png" type="BINDATA" />
+        <include name="IDR_PRODUCT_LOGO_128" file="../../../hex/browser/resources/hex_logo_128.png" type="BINDATA" />
         <!-- NOTE: product_logo_*.* files beyond what's listed above are
              referenced by installer code; don't remove them unless you know
              what you're doing! -->
         <if expr="is_macosx">
-          <include name="IDR_STATUS_TRAY_ICON" file="chromium/product_logo_22_mono.png" type="BINDATA" />
-          <include name="IDR_STATUS_TRAY_ICON_PRESSED" file="chromium/product_logo_22_mono_invert.png" type="BINDATA" />
+          <include name="IDR_STATUS_TRAY_ICON" file="../../../hex/browser/resources/hex_logo_22_mono.png" type="BINDATA" />
+          <include name="IDR_STATUS_TRAY_ICON_PRESSED" file="../../../hex/browser/resources/hex_logo_22_mono_invert.png" type="BINDATA" />
         </if>
         <if expr="not is_macosx">
-          <include name="IDR_STATUS_TRAY_ICON" file="chromium/product_logo_22.png" type="BINDATA" />
-          <include name="IDR_STATUS_TRAY_ICON_PRESSED" file="chromium/product_logo_22.png" type="BINDATA" />
+          <include name="IDR_STATUS_TRAY_ICON" file="../../../hex/browser/resources/hex_logo_22.png" type="BINDATA" />
+          <include name="IDR_STATUS_TRAY_ICON_PRESSED" file="../../../hex/browser/resources/hex_logo_22.png" type="BINDATA" />
         </if>
       </if>
       <if expr="is_win">
Index: app/theme/chromium/BRANDING
===================================================================
--- app/theme/chromium/BRANDING	(revision 220934)
+++ app/theme/chromium/BRANDING	(working copy)
@@ -1,5 +1,5 @@
 COMPANY_FULLNAME=The Chromium Authors
 COMPANY_SHORTNAME=The Chromium Authors
-PRODUCT_FULLNAME=Chromium
-PRODUCT_SHORTNAME=Chromium
+PRODUCT_FULLNAME=HeXium
+PRODUCT_SHORTNAME=HeXium
 COPYRIGHT=Copyright 2013 The Chromium Authors. All rights reserved.
Index: app/theme/theme_resources.grd
===================================================================
--- app/theme/theme_resources.grd	(revision 220934)
+++ app/theme/theme_resources.grd	(working copy)
@@ -647,12 +647,12 @@
       <structure type="chrome_scaled_image" name="IDR_PLUGINS_FAVICON" file="common/favicon_extensions.png" />
       <structure type="chrome_scaled_image" name="IDR_PRERENDER" file="prerender_succeed_icon.png" />
       <if expr="not pp_ifdef('_google_chrome')">
-        <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO" file="chromium/product_logo.png" />
-        <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO_16" file="chromium/product_logo_16.png" />
-        <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO_26" file="chromium/product_logo_26.png" />
-        <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO_32" file="chromium/product_logo_32.png" />
-        <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO_NAME_48" file="chromium/product_logo_name_48.png" />
-        <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO_WHITE" file="chromium/product_logo_white.png" />
+        <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO" file="../../../../hex/browser/resources/hex_logo_name.png" />
+        <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO_16" file="../../../../hex/browser/resources/hex_logo_16.png" />
+        <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO_26" file="../../../../hex/browser/resources/hex_logo_26.png" />
+        <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO_32" file="../../../../hex/browser/resources/hex_logo_32.png" />
+        <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO_NAME_48" file="../../../../hex/browser/resources/hex_logo_name_48.png" />
+        <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO_WHITE" file="../../../../hex/browser/resources/hex_logo_name_white.png" />
       </if>
       <if expr="pp_ifdef('_google_chrome')">
         <structure type="chrome_scaled_image" name="IDR_PRODUCT_LOGO" file="google_chrome/product_logo.png" />
Index: browser/chrome_browser_main_mac.mm
===================================================================
--- browser/chrome_browser_main_mac.mm	(revision 220934)
+++ browser/chrome_browser_main_mac.mm	(working copy)
@@ -187,6 +187,15 @@
   }
 
   RecordCatSixtyFour();
+
+  // For heXium
+  if (!parsed_command_line().HasSwitch(switches::kCurrentWorkingDirectory)) {
+    NSString* path = [[NSBundle mainBundle] resourcePath];
+    std::string path_str = [path UTF8String];
+    CommandLine::ForCurrentProcess()->AppendSwitchASCII(
+        switches::kCurrentWorkingDirectory, path_str);
+    //[[NSFileManager defaultManager] changeCurrentDirectoryPath:path];
+  }
 }
 
 void ChromeBrowserMainPartsMac::PreMainMessageLoopStart() {
Index: browser/chrome_browser_main_win.cc
===================================================================
--- browser/chrome_browser_main_win.cc	(revision 220934)
+++ browser/chrome_browser_main_win.cc	(working copy)
@@ -161,6 +161,15 @@
       }
     }
   }
+
+  // For heXium
+  if (!CommandLine::ForCurrentProcess()->HasSwitch(
+      switches::kCurrentWorkingDirectory)) {
+    char path[MAX_PATH];
+    GetCurrentDirectoryA(MAX_PATH, path);
+    CommandLine::ForCurrentProcess()->AppendSwitchASCII(
+        switches::kCurrentWorkingDirectory, path);
+  }
 }
 
 ChromeBrowserMainPartsWin::~ChromeBrowserMainPartsWin() {
Index: browser/resources/help/help.html
===================================================================
--- browser/resources/help/help.html	(revision 220934)
+++ browser/resources/help/help.html	(working copy)
@@ -49,6 +49,9 @@
     </div>
     <div id="version-container">
       <div i18n-content="browserVersion" dir="ltr"></div>
+      <!-- For heXium -->
+      <div i18n-content="hexVersion" dir="ltr"></div>
+      <div i18n-content="nodeVersion" dir="ltr"></div>
 <if expr="pp_ifdef('chromeos')">
       <div>
         <span i18n-content="platform"></span>&nbsp;<span id="os-version"></span>
Index: browser/ui/startup/bad_flags_prompt.cc
===================================================================
--- browser/ui/startup/bad_flags_prompt.cc	(revision 220934)
+++ browser/ui/startup/bad_flags_prompt.cc	(working copy)
@@ -28,7 +28,8 @@
   static const char* kBadFlags[] = {
     // These imply disabling the sandbox.
     switches::kSingleProcess,
-    switches::kNoSandbox,
+    // For heXium, we must disable sandbox.
+    //switches::kNoSandbox,
     switches::kDisableWebSecurity,
     // Browser plugin is dangerous on regular pages because it breaks the Same
     // Origin Policy.
Index: browser/ui/webui/help/help_handler.cc
===================================================================
--- browser/ui/webui/help/help_handler.cc	(revision 220934)
+++ browser/ui/webui/help/help_handler.cc	(working copy)
@@ -53,6 +53,11 @@
 #include "content/public/browser/browser_thread.h"
 #endif
 
+// For heXium
+#include "base/strings/stringprintf.h"
+#include "hex/include/hex_version.h"
+#include "third_party/node/src/node_version.h"
+
 using base::ListValue;
 using content::BrowserThread;
 
@@ -239,6 +244,15 @@
       l10n_util::GetStringFUTF16(IDS_ABOUT_PRODUCT_VERSION,
                                  BuildBrowserVersionString()));
 
+  // For heXium
+  source->AddString("hexVersion", base::StringPrintf("heX version: %d.%d.%d-%s",
+                                                     HEX_VERSION_MAJOR,
+                                                     HEX_VERSION_MINOR,
+                                                     HEX_VERSION_BUILD,
+                                                     HEX_REVISION));
+  source->AddString("nodeVersion", base::StringPrintf("Node.js version: %s",
+                                                      NODE_VERSION_STRING));
+
   string16 license = l10n_util::GetStringFUTF16(
       IDS_ABOUT_VERSION_LICENSE,
       ASCIIToUTF16(chrome::kChromiumProjectURL),
Index: chrome.gyp
===================================================================
--- chrome.gyp	(revision 220934)
+++ chrome.gyp	(working copy)
@@ -400,7 +400,8 @@
           ],
           'xcode_settings': {
             'CHROMIUM_BUNDLE_ID': '<(mac_bundle_id)',
-            'CHROMIUM_SHORT_NAME': '<(branding)',
+            # For heXium
+            'CHROMIUM_SHORT_NAME': 'HeXium',
             'CHROMIUM_STRIP_SAVE_FILE': 'app/app.saves',
             'INFOPLIST_FILE': 'app/helper-Info.plist',
           },
@@ -436,6 +437,17 @@
               ],
             },
             {
+              # For heXium
+              'postbuild_name': 'Fix heX Link',
+              'action': [
+                'install_name_tool',
+                '-change',
+                '@executable_path/../Versions/<(version_full)/<(mac_product_name) Framework.framework/Libraries/libhex.dylib',
+                '@executable_path/../../../<(mac_product_name) Framework.framework/Libraries/libhex.dylib',
+                '${BUILT_PRODUCTS_DIR}/${EXECUTABLE_PATH}'
+              ],
+            },
+            {
               # Modify the Info.plist as needed.  The script explains why this
               # is needed.  This is also done in the chrome and chrome_dll
               # targets.  In this case, --breakpad=0, --keystone=0, and --scm=0
Index: chrome_dll.gypi
===================================================================
--- chrome_dll.gypi	(revision 220934)
+++ chrome_dll.gypi	(working copy)
@@ -82,6 +82,8 @@
             '<@(chromium_browser_dependencies)',
             '../content/content.gyp:content_app_browser',
             'app/policy/cloud_policy_codegen.gyp:policy',
+            # For heXium
+            '<(DEPTH)/hex/hex.gyp:hex',
           ],
           'conditions': [
             ['use_aura==1', {
@@ -274,6 +276,21 @@
                 },
               ],
               'conditions': [
+                # For heXium
+                ['OS=="mac"', {
+                  'postbuilds': [
+                    {
+                      # For heXium, Fix V8, UV, Node Link
+                      'postbuild_name': 'Fix V8, UV, Node Link',
+                      'action': [
+                        'install_name_tool',
+                        '-add_rpath',
+                        '@loader_path/./Libraries',
+                        '${BUILT_PRODUCTS_DIR}/${EXECUTABLE_PATH}'
+                      ],
+                    },
+                  ],
+                }],
                 ['mac_breakpad_compiled_in==1', {
                   'dependencies': [
                     '../breakpad/breakpad.gyp:breakpad',
Index: chrome_dll_bundle.gypi
===================================================================
--- chrome_dll_bundle.gypi	(revision 220934)
+++ chrome_dll_bundle.gypi	(working copy)
@@ -143,6 +143,10 @@
       'destination': '<(PRODUCT_DIR)/$(CONTENTS_FOLDER_PATH)/Libraries',
       'files': [
         '<(PRODUCT_DIR)/ffmpegsumo.so',
+        '<(PRODUCT_DIR)/libhex.dylib',
+        '<(PRODUCT_DIR)/libv8.dylib',
+        '<(PRODUCT_DIR)/libnode.dylib',
+        '<(PRODUCT_DIR)/libuv.dylib',
       ],
     },
     {
Index: chrome_exe.gypi
===================================================================
--- chrome_exe.gypi	(revision 220934)
+++ chrome_exe.gypi	(working copy)
@@ -43,6 +43,12 @@
         'CHROMIUM_STRIP_SAVE_FILE': 'app/app.saves',
         'INFOPLIST_FILE': 'app/app-Info.plist',
       },
+      # For heX
+      'msvs_settings': {
+        'VCLinkerTool': {
+          'OutputFile': '$(OutDir)hexium$(TargetExt)',
+        },
+      },
       'conditions': [
         ['component == "shared_library"', {
           'variables': {
@@ -136,7 +142,8 @@
                     ['branding=="Chrome"', {
                       'files': ['app/theme/google_chrome/product_logo_48.png']
                     }, { # else: 'branding!="Chrome"
-                      'files': ['app/theme/chromium/product_logo_48.png']
+                      # For heXium
+                      'files': ['../hex/browser/resources/hex_logo_48.png']
                     }],
                   ],
                 },
@@ -189,7 +196,7 @@
               ],
             }, {  # else: 'branding!="Chrome"
               'mac_bundle_resources': [
-                'app/theme/chromium/app.icns',
+                '../hex/browser/resources/hex.icns',
                 'app/theme/chromium/document.icns',
                 'browser/ui/cocoa/applescript/scripting.sdef',
               ],
@@ -253,7 +260,8 @@
             # as settings on the target.
             'CHROMIUM_BUNDLE_ID': '<(mac_bundle_id)',
             'CHROMIUM_CREATOR': '<(mac_creator)',
-            'CHROMIUM_SHORT_NAME': '<(branding)',
+            # For heXium
+            'CHROMIUM_SHORT_NAME': 'HeXium',
           },
           'dependencies': [
             'helper_app',
Index: chrome_renderer.gypi
===================================================================
--- chrome_renderer.gypi	(revision 220934)
+++ chrome_renderer.gypi	(working copy)
@@ -33,12 +33,27 @@
         '../webkit/renderer/webkit_renderer.gyp:webkit_renderer',
         '../webkit/support/webkit_support.gyp:glue',
         '../webkit/webkit_resources.gyp:webkit_resources',
+        # For heXium
+        '<(DEPTH)/third_party/node/node.gyp:node',
+        '<(DEPTH)/third_party/node/deps/uv/uv.gyp:libuv',
       ],
       'include_dirs': [
         '..',
         # Needed by chrome_content_renderer_client.cc.
         '<(SHARED_INTERMEDIATE_DIR)',
+        # For heXium
+        '<(DEPTH)/hex/src',
+        '<(DEPTH)/third_party/node/src',
+        '<(DEPTH)/third_party/node/deps/uv/include',
       ],
+      # For heXium
+      'msvs_disabled_warnings': [
+        4251,
+        4996,
+        4390,
+        4800,
+        4005,
+      ],
       'defines': [
         '<@(nacl_defines)',
       ],
@@ -338,6 +353,9 @@
         'renderer/web_apps.h',
         'renderer/webview_color_overlay.cc',
         'renderer/webview_color_overlay.h',
+        # For heXium
+        '<(DEPTH)/hex/src/init_node.cc',
+        '<(DEPTH)/hex/src/init_node.h',
       ],
       'conditions': [
         ['disable_nacl!=1', {
Index: common/url_constants.cc
===================================================================
--- common/url_constants.cc	(revision 220934)
+++ common/url_constants.cc	(working copy)
@@ -365,7 +365,8 @@
     "https://support.google.com/chromeos/?p=help&ctx=settings";
 #endif  // defined(OFFICIAL_BUILD
 #else
-    "https://support.google.com/chrome/?p=help&ctx=settings";
+    // For heXium
+    "http://hex.youdao.com/hexium";
 #endif  // defined(OS_CHROMEOS)
 
 #if defined(OS_CHROMEOS)
Index: renderer/chrome_content_renderer_client.cc
===================================================================
--- renderer/chrome_content_renderer_client.cc	(revision 220934)
+++ renderer/chrome_content_renderer_client.cc	(working copy)
@@ -107,6 +107,16 @@
 #include "chrome/renderer/spellchecker/spellcheck_provider.h"
 #endif
 
+// For heXium
+#if !defined(HEXCLIENT)
+#include "hex/src/init_node.h"
+#include "third_party/node/src/node.h"
+#include "third_party/node/deps/uv/include/uv.h"
+#include "hex/src/hex_object.h"
+#include "hex/src/hex_switches_port.h"
+#include "hex/src/hex_use_node.h"
+#endif  // !HEXCLIENT
+
 using autofill::AutofillAgent;
 using autofill::PasswordAutofillAgent;
 using autofill::PasswordGenerationManager;
@@ -350,6 +360,11 @@
 
 void ChromeContentRendererClient::RenderViewCreated(
     content::RenderView* render_view) {
+  // For heXium, initialize Node.js.
+#if !defined(HEXCLIENT)
+  INIT_NODE();
+#endif  // !HEXCLIENT
+
   ContentSettingsObserver* content_settings =
       new ContentSettingsObserver(render_view);
   if (chrome_observer_.get()) {
